### PR TITLE
rgw: support new zone pools

### LINKED ISSUE #14708: RGW: shared pools: extra `<storename>.rgw.meta` pool created
One issue I (Travis) noticed [with shared pools] is that an extra pool is created automatically. For example, if I create an object store named `store-d`, a pool will appear called `store-d.rgw.meta`. This could be a result of Squid, or maybe some other issue since I don't recall seeing this extra pool before.

_Originally posted by @travisn in https://github.com/rook/rook/issues/14588#issuecomment-2334832033_

Let's look into this to see if we can get this defined in the store/zone spec to go into an existing pool. Probably the existing `.rgw.meta` pool is the right place.

### PR DESCRIPTION
**Pull request by arttor**
closes #14708

### CODE REVIEW

File: pkg/operator/ceph/object/objectstore.go
Comment (travisn) [inline pkg/operator/ceph/object/objectstore.go:919]: If we try to set these new properties in v18, does it fail? I assume that's why you're checking if the properties exist like this.

File: pkg/operator/ceph/object/objectstore.go
Comment (travisn) [inline pkg/operator/ceph/object/objectstore.go:926]: An alternative to the "IfExists" approach, is to check the ceph version. Something like this:
```
if cephVersion.IsAtLeast(cephver.CephVersion{Major: 19, Minor: 0, Extra: 0}) {
  // add new properties
}
```
But if we can just check if the property exists, that may certainly keep things simpler in the long term when other properties are added. 

During an upgraded cluster (e.g. from v18 --> v19), the new properties will be found to exist and then we'll add them? Just wanted to confirm if you already tested that scenario.

Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:926]: I think that Ceph version check approach is better for code readability. But, as you said, with "IfExists" approach we can just add any new pools and don't care about ceph version. 
I personally more like "IfExists" approach because it makes RGW responsible for pools structure and rook just tries to apply namespaces to all existing pools. 
I can also suggest iterating through all root string fields ending with `_pool`  and log warning if pool is unknown. Or try to apply some namespace anyway. What do you think?


I've upgaded my local rook-ceph from `v18.something` to `v19.1.1-0`. All daemon pods been restared and using v19 images, but when i run `radosgw-admin zone get` i don't see any new pools in json after update:
```json
{
    "id": "e522c205-dbd2-43e1-8ef8-a81a55d44ef1",
    "name": "pp-store",
    "domain_root": "meta-pool:pp-store.meta.root",
    "control_pool": "meta-pool:pp-store.control",
    "gc_pool": "meta-pool:pp-store.log.gc",
    "lc_pool": "meta-pool:pp-store.log.lc",
    "log_pool": "meta-pool:pp-store.log",
    "intent_log_pool": "meta-pool:pp-store.log.intent",
    "usage_log_pool": "meta-pool:pp-store.log.usage",
    "roles_pool": "meta-pool:pp-store.meta.roles",
    "reshard_pool": "meta-pool:pp-store.log.reshard",
    "user_keys_pool": "meta-pool:pp-store.meta.users.keys",
    "user_email_pool": "meta-pool:pp-store.meta.users.email",
    "user_swift_pool": "meta-pool:pp-store.meta.users.swift",
    "user_uid_pool": "meta-pool:pp-store.meta.users.uid",
    "otp_pool": "meta-pool:pp-store.otp",
    "system_key": {
        "access_key": "",
        "secret_key": ""
    },
    "placement_pools": [...],
    "realm_id": "ceb20ae5-3e69-40c4-af9f-fc9bab7904db",
    "notif_pool": "meta-pool:pp-store.log.notif"
}
```

File: pkg/operator/ceph/object/objectstore.go
Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:919]: Correct, if pool is not presented in source json then we just ignore it.

File: pkg/operator/ceph/object/objectstore.go
Comment (travisn) [reply pkg/operator/ceph/object/objectstore.go:926]: > I can also suggest iterating through all root string fields ending with _pool and log warning if pool is unknown. Or try to apply some namespace anyway. What do you think?

That sounds great to log a warning if we find an extra pool, until we are more confident with that approach.

> I've upgaded my local rook-ceph from v18.something to v19.1.1-0. All daemon pods been restared and using v19 images, but when i run radosgw-admin zone get i don't see any new pools in json after update:

This is what I worried about, the schema may not be updated until some zone update is triggered. So that means we may need to use the approach to check the ceph version instead of the "IfExists" approach.

File: pkg/operator/ceph/object/json_helpers.go
Comment (BlaineEXE) [inline pkg/operator/ceph/object/json_helpers.go:94]: This function looks to be ~90% copy-pasted from `setObjProperty()`. Is there a good way to abstract out the main logic into a helper, that can be used from both `setObjProperty()` and `setObjPropertyIfExists()`? Or alternately, could `setObjPropertyIfExists()` be called by `setObjProperty()`? 

I think this will be best for long term maintenance. If there are any logic corner cases we need to add, it will help to be able to add/fix the logic in only one function.

File: pkg/operator/ceph/object/objectstore.go
Comment (BlaineEXE) [reply pkg/operator/ceph/object/objectstore.go:926]: > That sounds great to log a warning if we find an extra pool, until we are more confident with that approach.

I really like this idea as well. 

-----

Conceptually, I much prefer the `IfExists` approach, but it does seem like we have an edge condition on upgrade we need figure out how to deal with properly. 

If we upgrade from v17/v18 -> v19, my expectation would be that RGW (upon upgrade) is going to create a new pool before we Rook has a chance to see if a new pool needs configured. I'm wondering what bad would happen if the RGW did that and started putting stuff into that pool before Rook has a chance to configure it. But it also seems like my expectation is not what happens IRL, which is also concerning.

I think it's best to ask @cbodley what his recommendation is for handling the upgrade case when new fields are introduced to the zone config. How does RGW operate on upgrade when the zone spec is missing configs for default pools? Does RGW create the new default pools and use them implicitly, or does it just not use the new default pools? Is there a good (preferably idempotent) way of checking to see if the zone config is missing fields it needs after an RGW upgrade? Are there other things around this we're missing?

I think the best-case scenario for Rook would be if RGW doesn't init new default pools on upgrade and we have a recommendation for detecting what pools RGW wants to be configured that we can use on an if-exists basis.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/objectstore.go
Comment (cbodley) [reply pkg/operator/ceph/object/objectstore.go:926]: > I've upgaded my local rook-ceph from `v18.something` to `v19.1.1-0`. All daemon pods been restared and using v19 images, but when i run `radosgw-admin zone get` i don't see any new pools in json after update:

@arttor that makes it sound like the `radosgw-admin` binary was still v18

Comment (cbodley) [reply pkg/operator/ceph/object/objectstore.go:926]: @BlaineEXE the upgrade case is handled by rgw during deserialization of `struct RGWZoneParams`: https://github.com/ceph/ceph/blob/squid/src/rgw/driver/rados/rgw_zone.h#L258-L265

if it finds an earlier version of the encoding (ie written by a reef rgw), it fills in missing pool names with their default values. after decode, rgw may go on to create and write into those pools

if rook wants to prevent this, it would have to use an upgraded `radosgw-admin zone set` command to update those pool names before any other radosgws/radosgw-admin commands get upgraded

reef rgws would still be able to read that updated encoding and, as long as they don't overwrite RGWZoneParams with their old encoding version, they'd see the updated pool names after upgrading to squid

Comment (cbodley) [reply pkg/operator/ceph/object/objectstore.go:926]: > I'm wondering what bad would happen if the RGW did that and started putting stuff into that pool before Rook has a chance to configure it.

if rook changes pool names after rgw has already created and written to the default pool names, that would essentially look like data loss

Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:926]: it is probably something wrong with my setup. even when i try to create a new rgw, i don't see new pools in zone json. RGW is using this image [`quay.io/ceph/ceph@sha256:096f9e9a97636092e354ecc1a3fd411579d3c6adbbe1a195ad58ea84ad910de9
`](https://quay.io/repository/ceph/ceph/manifest/sha256:096f9e9a97636092e354ecc1a3fd411579d3c6adbbe1a195ad58ea84ad910de9)

Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:926]: oh, yea. You are correct, i was using old `radosgw-admin` 🤦 . After toolbox update i can see new pools. 
For existing rgw with shared pools, new pools are using `<instance_name>.rgw.meta` pool:
```json
{
    "notif_pool": "meta-pool:pp-store.log.notif", // existing is using shared metadata pool "meta-pool"
    "topics_pool": "pp-store.rgw.meta:topics", // new pool using default "pp-store.rgw.meta"
    "account_pool": "pp-store.rgw.meta:accounts", 
    "group_pool": "pp-store.rgw.meta:groups",
...
}
```

Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:926]: > if rook wants to prevent this, it would have to use an upgraded `radosgw-admin zone set` command to update those pool names before any other radosgws/radosgw-admin commands get upgraded

@BlaineEXE how can we achive this? It would be very easy if we could add new pools with old `radosgw-admin`. Or it is ok to ask users to upgrade rook to the latest version before upgrading Ceph? If it is ok, then we can just always force set new pools to zone and it should work.

Comment (BlaineEXE) [reply pkg/operator/ceph/object/objectstore.go:926]: Since we know about the pools, can we attempt to always set them regardless of the Ceph version? That way, they will be ready on any upgrade.

Comment (BlaineEXE) [reply pkg/operator/ceph/object/objectstore.go:926]: > that makes it sound like the radosgw-admin binary was still v18

That's a good point, and this is accurate. Because Ceph v19 is still pre-release, Rook is using the latest v18 image for its base image. That means that the Ceph CLI and all tools a re v18 tools, including radosgw-admin.

**Review by arttor — COMMENTED** [review]
@BlaineEXE @travisn 

i've done some more testing locally. Tried 4 possible combinations with v18/v19 ceph-daemons and rook-ceph-dependencies. I was not able to find any errors in rgw or operator logs for a simple scenarios like put/get/head bucket and object. I've also tested bucket notifications for v19 rook and v18 rgw combination and it worked.

I propose to use `setIfExistApproach` approach for zone pools. Otherwise there will be unnecesary attempt of zone update for rook v18 case because new pools won't be saved to zone json.

File: pkg/operator/ceph/object/json_helpers.go
Comment (BlaineEXE) [inline pkg/operator/ceph/object/json_helpers.go:76]: I would expect a bool called `shouldExist` to cause the function to return an error if the field doesn't exist. Is there a better naming for this? Maybe `setIfExists`?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Really just a small nit around naming as far as the code is written.

File: pkg/operator/ceph/object/json_helpers.go
Comment (BlaineEXE) [reply pkg/operator/ceph/object/json_helpers.go:76]: @arttor this is also bringing up some related questions for me, and I'd like to clarify my understanding.

From reading the overall code, it seems like there is no error returned if the field Rook tries to set doesn't exist. Maybe I am misreading the code; if so let me know. If I'm reading correctly, is this for a specific reason?

IMO, it's probably safest for users for Rook to return an error if it tries to set a field using `setObjProperty()` if the field doesn't exist. Presumably, that case would happen if (1) RGW removes a field in the future, which we want to make sure we handle properly in code if needed, or (2) a dev makes a typo/mistake in a future Rook update.

Being able to catch (2) would be good, though not critical. Being able to catch (1) could be good to help make sure Rook's handling of RGW zone configs is always aware of RGW changes. I don't expect (1) to happen often, but catching it could be important.

Comment (arttor) [reply pkg/operator/ceph/object/json_helpers.go]: done. now it retruns error if pool not exists and the error is logged with INFO level.

File: pkg/operator/ceph/object/json_helpers.go
Comment (BlaineEXE) [inline pkg/operator/ceph/object/json_helpers.go:61]: Are these switched from what they should be? 

I think `shouldExist=false` when using `setObjPropertyIfExists()` and `shouldExist=true` when using `setObjProperty()` makes sense right?

```suggestion
	return setObjPropertyCommon(true, obj, val, path...)
}

// setObjPropertyIfExists - helper function to manipulate JSON Objects.
// sets new value to json object nested field only if it is already exists in json and returns previous value.
func setObjPropertyIfExists[T string | []string | map[string]interface{} | []interface{}](obj map[string]interface{}, val T, path ...string) (T, error) {
	return setObjPropertyCommon(false, obj, val, path...)
}
```

File: pkg/operator/ceph/object/objectstore.go
Comment (BlaineEXE) [inline pkg/operator/ceph/object/objectstore.go:926]: For the pools added here, I think `setObjPropertyIfExists()` makes sense, but for existing pools, should we use `setObjProperty()` so that Rook can error out if the property is missing?

File: pkg/operator/ceph/object/json_helpers.go
Comment (arttor) [reply pkg/operator/ceph/object/json_helpers.go:61]: no, it is correct. `setObjPropertyIfExists()` means that `shouldExist=true`. Is it makes sense to change namig?

File: pkg/operator/ceph/object/objectstore.go
Comment (arttor) [reply pkg/operator/ceph/object/objectstore.go:926]: i think that we already have described behaviour with `setObjPropertyIfExists()`. it will return error if pool is missing.
Old pools should always be presented. Even it is missing in ceph, `radosgw-admin` will set default values to json anyway. If it will be missing in json then we will see warning in rook logs.

File: pkg/operator/ceph/object/json_helpers.go
Comment (BlaineEXE) [reply pkg/operator/ceph/object/json_helpers.go:61]: I think the confusion I have is the use of `shouldExist`. I am interpreting this to be the same as `mustExist`, effectively `failIfNotExist`.

Perhaps the parameter `shouldExist` could be renamed to `allowNotExists`?

Comment (arttor) [reply pkg/operator/ceph/object/json_helpers.go:61]: You interpreting it correctly. It acts like `mustExist` and will return error if property is not exists

Comment (BlaineEXE) [reply pkg/operator/ceph/object/json_helpers.go:61]: I read through the unit tests, and I think I understand where my confusion is coming from.

In the unit tests, I see that `setObjProperty()` will inject a property into the json if the property doesn't already exist. Based on the name, I have been assuming that the `setObjProperty()` function would error out if the property it is operating on doesn't exist because I interpret "set" to be equivalent to "update".

I also see that `setObjPropertyIfExists()` returns an error if the property doesn't exist. Based on the name, I have been assuming that the `setObjPropertyIfExists()` doesn't need to return an error if the property is non-existent because of "if exists".

I have some follow-ups thoughts to try to get/create clarity:

Does Rook need to be able to add fields to the json struct for Ceph? 
If yes, renaming `setObjProperty()` to `setOrAddObjProperty()` (or `addOrUpdateObjProperty()`) could be a good clarifying change. Following from that, `setObjPropertyIfExists()` might be better named as `setObjProperty()` (or `updateObjProperty()`).

If Rook doesn't need to add fields to the json struct, maybe that feature could be removed from the methods? This case probably needs more discussion.

Comment (arttor) [reply pkg/operator/ceph/object/json_helpers.go:61]: Thanks for clarification, now i also see the problem. You are right, we use this method only for zone and zonegroup JSONs. And these JSONs are returned from `radosgw-admin` and should always have all properties. This means that at least for now we can use `setObjPropertyIfExists()` for all cases.
I will remove `setObjProperty()` method and rename `setObjPropertyIfExists()` to `updateObjProperty()`.

Comment (arttor) [reply pkg/operator/ceph/object/json_helpers.go]: done

**Review by BlaineEXE — APPROVED** [review]
I think this looks good. Thanks @arttor for taking the time to do this and working to make sure the code is as readable and maintainable as possible going forward!

I'll let @travisn have another look if he wants before merging

**Review by travisn — APPROVED** [review]

**Pull request merged**
